packages/htmlbars-util/lib/template-utils.js

```diff
@@ -36,25 +37,30 @@ Block.prototype.invoke = function(env, blockArguments, self, renderNode, parentS
   if (renderNode.lastResult) {
     renderNode.lastResult.revalidateWith(env, undefined, self, blockArguments, visitor);
   } else {
-    let options = { renderState: new RenderState(renderNode) };
-    let { render, template, blockOptions: { scope } } = this;
-    let shadowScope = scope ? env.hooks.createChildScope(scope) : env.hooks.createFreshScope();
-
-    env.hooks.bindShadowScope(env, parentScope, shadowScope, this.blockOptions.options);
+    this._firstRender(env, blockArguments, self, renderNode, parentScope);
+  }
+};
 
-    if (self !== undefined) {
-      env.hooks.bindSelf(env, shadowScope, self);
-    } else if (this.blockOptions.self !== undefined) {
-      env.hooks.bindSelf(env, shadowScope, this.blockOptions.self);
-    }
+Block.prototype._firstRender = function(env, blockArguments, self, renderNode, parentScope) {
+  let options = { renderState: new RenderState(renderNode) };
+  let { render, template, blockOptions: { scope } } = this;
+  let shadowScope = scope ? env.hooks.createChildScope(scope) : env.hooks.createFreshScope();
 
-    bindBlocks(env, shadowScope, this.blockOptions.yieldTo);
+  env.hooks.bindShadowScope(env, parentScope, shadowScope, this.blockOptions.options);
 
-    renderAndCleanup(renderNode, env, options, null, function() {
-      options.renderState.morphToClear = null;
-      render(template, env, shadowScope, { renderNode, blockArguments });
-    });
+  if (self !== undefined) {
+    env.hooks.bindSelf(env, shadowScope, self);
+  } else if (this.blockOptions.self !== undefined) {
+    env.hooks.bindSelf(env, shadowScope, this.blockOptions.self);
   }
+
+  bindBlocks(env, shadowScope, this.blockOptions.yieldTo);
+
+  renderAndCleanup(renderNode, env, options, null, function() {
```

Comment:
is it possible for us to not use a closure here, rather just pass in some task ?